sdk_extensions/zpages/src/main/java/io/opentelemetry/sdk/extensions/zpages/TracezSpanProcessor.java

```diff
@@ -0,0 +1,193 @@
+/*
+ * Copyright 2020, OpenTelemetry Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.opentelemetry.sdk.extensions.zpages;
+
+import com.google.common.annotations.VisibleForTesting;
+import io.opentelemetry.sdk.common.export.ConfigBuilder;
+import io.opentelemetry.sdk.trace.ReadableSpan;
+import io.opentelemetry.sdk.trace.SpanProcessor;
+import io.opentelemetry.trace.SpanId;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.Map;
+import javax.annotation.concurrent.ThreadSafe;
+
+/**
+ * A {@link SpanProcessor} implementation for the traceZ zPage.
+ *
+ * <p>Configuration options for {@link io.opentelemetry.sdk.extensions.zpages.TracezSpanProcessor}
+ * can be read from system properties, environment variables, or {@link java.util.Properties}
+ * objects.
+ *
+ * <p>For system properties and {@link java.util.Properties} objects, {@link
+ * io.opentelemetry.sdk.extensions.zpages.TracezSpanProcessor} will look for the following names:
+ *
+ * <ul>
+ *   <li>{@code otel.ssp.export.sampled}: sets whether only sampled spans should be exported.
+ * </ul>
+ *
+ * <p>For environment variables, {@link io.opentelemetry.sdk.extensions.zpages.TracezSpanProcessor}
+ * will look for the following names:
+ *
+ * <ul>
+ *   <li>{@code OTEL_SSP_EXPORT_SAMPLED}: sets whether only sampled spans should be exported.
+ * </ul>
+ */
+@ThreadSafe
+public final class TracezSpanProcessor implements SpanProcessor {
+  private final Map<SpanId, ReadableSpan> runningSpanCache;
+  private final Map<SpanId, ReadableSpan> completedSpanCache;
```

Comment:
I see there is mention about capping the size of this map, but in general I'm sort of concerned about the design of the class if it's based on keeping a map of all completed spans. Can we incrementally update data structures in this class instead? Keeping running spans as a map is ok I think since it should be bounded and avoids having to both incrementally update and revert structures, but if I'm not missing it, it seems that all the structures computed from this map can be inlined and updated incrementally instead?

Of course if going with incremental, the comment about Immutable goes away :)

Reply:
I've updated the `completedSpanCache` to limit the number of latency and error spans. I'm not quite sure what you mean by updating the cache incrementally, but I hope the new changes address some of your concerns.

Reply:
So for example, we have a method `getSpanNames`, instead of iterating over running /completed spans, the processor can be something like

```
TracezSpanProcessor

private final Set<String> spanNames;

onStart(span) {
  spanNames.add(span.name);
}

onEnd(span) {
  spanNames.add(span.name);
}
```

And `getSpanNames` can just return that directly. From what I can tell, the information is mostly sets of names and counts, so I think we could do a similar approach for all of them (I didn't do a deep dive though so let me know if I missed something).

With the cap on the cache, things are still much safer, but if we keep track of the metrics themselves without storing spans, we wouldn't even need to worry about it, and I guess have better metrics? What do you think?

Reply:
Ah, I notice with the new code, actually names is using this pattern. Can't we apply something similar to all the metrics?

Reply:
This makes sense for span names, so I'll go ahead and modify the implementation. For other metrics (counts in particular), what exactly did you have in mind? I'm not opposed to maintaining separate list for spans and counts, but I feel like this results in a bunch of repeated information. In addition, a few of the functions in `TracezDataAggregator` filter by span name or latency. Should a separate list be maintained for each of these functions?